### PR TITLE
onChanged callback object content clarification

### LINKED ISSUE #31203: Conflicting Information regarding storage onChanged
### MDN URL

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/onChanged

### What specific section or headline is this issue about?

changes

### What information was incorrect, unhelpful, or incomplete?

Due to extension bug reports, I had to test the output of `onChanged` and I found out that there are conflicting information on MDN which led to the issue.

> [storage.onChanged](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/onChanged#changes)
> 
> `object`. Object describing the change. This object contains properties for all the keys in the storage area included in the [storageArea.set](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/set) call, even if key values are unchanged.

> [storage.StorageArea.onChanged](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/onChanged#changes)
> 
> `object`. Object describing the change. This contains one property for each key that changed. The name of the property is the name of the key that changed, and its value is a [storage.StorageChange](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageChange) object describing the change to that item.

First one says: "all the keys in the storage area, even if key values are unchanged"
Second one says: "one property for each key that changed."

After further testing both APIs `storage.onChanged` & `storage.sync.onChanged`:

- Firefox `onChanged` returns all the keys in the storage area, even if all values are unchanged
- Chrome `onChanged` returns only the changed values (event doesn't fire when there is no change)

### What did you expect to see?

no conflicting information

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

See also:
- [Bug 1833153: storage.sync should not emit onChanged events for calls that are not resulting into changes to the store data](https://bugzilla.mozilla.org/show_bug.cgi?id=1833153)

### PR DESCRIPTION
**Pull request by rebloor**
### Description

Clarifies that the `storage.onChanged` callback triggered as a result of `storageArea.set` only returns changed details.

### Related issues and pull requests

Fixes #31203

### CODE REVIEW

**Comment by erosman** [comment]
The issue had to be updated.
Please hold the PR.

Sorry for the inaccurate test results.

**Comment by rebloor** [comment]
@erosman updated based on your issue revisions

**Comment by erosman** [comment]
Just a note which may save some developers some time. 
Feel free to include something about it or not.

While Firefox storage `onChanged` (currently, see [Bug 1833153](https://bugzilla.mozilla.org/show_bug.cgi?id=1833153)) returns all the values and Chrome (currently) only the changed values, the target storage area still contains all the values e.g.

```js
// same for browser and chrome APIs
browser.storage.onChanged.addListener()
browser.storage.sync.onChanged.addListener()

// the following contains all the values passed to the storage sync, changed and unchanged
browser.storage.sync.get()
```

**Comment by rebloor** [comment]
@erosman sorry, I'm unclear what the note about `get` would achieve if the developers were trying to observe changed keys.

**Comment by erosman** [comment]
> @erosman sorry, I'm unclear what the note about `get` would achieve if the developers were trying to observe changed keys.


When developing a cross-browser-compatible extension, API inconsistency is a major concern. [1]
What I came up with in this case was to use `onChanged` as a trigger and then get all the values from `storage.sync.get()`, the same way Firefox returns `onChanged`.
While that is not necessary in all the cases, it is helpful in some cases (especially when a large values has to be segmented). [2]

I don't know if it is something worth mentioning or not, and feel free to disregard it.

Please also note 
1. [Inconsistency: storage onChanged](https://github.com/w3c/webextensions/issues/511)
2. [Proposal: Increase maximum item size in Storage sync quotas](https://github.com/w3c/webextensions/issues/510)

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 1]
```diff
@@ -7,9 +7,9 @@ browser-compat: webextensions.api.storage.onChanged
 
 {{AddonSidebar()}}
 
-Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area.
+Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area, returning details of the changed keys. A callback is received only when there are changes to the underlying data.
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:10]:
```suggestion
Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area, returning details of the changed keys. A callback is called only when there are changes to the underlying data.
```

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 2]
```diff
@@ -7,9 +7,9 @@ browser-compat: webextensions.api.storage.onChanged
...
 
-When this event is triggered by {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, it's possible to receive a callback when there is no change to the underlying data. Also, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. The extension can determine the changes that occurred by examining the content of the `changes` argument received by the `onChanged` listeners.
+> **Note:** In Firefox, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. Also, receiving a callback is possible when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:12]:
```suggestion
> **Note:** In Firefox, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. Also, a callback may be invoked when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```

File: files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md [hunk 3]
```diff
@@ -7,7 +7,9 @@ browser-compat: webextensions.api.storage.StorageArea.onChanged
 
 {{AddonSidebar()}}
 
-Fires when one or more items in a storage area change. Compared to {{WebExtAPIRef("storage.onChanged")}}, this event enables you to listen for changes in one of the storage areas: `local`, `managed`, `session`, and `sync`.
+Fires when one or more items in a storage area change, returning details for the keys that changed. Compared to {{WebExtAPIRef("storage.onChanged")}}, this event enables you to listen for changes in one of the storage areas: `local`, `managed`, `session`, and `sync`.
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md:10]: So, Firefox can pass changes for multiple areas, but everyone else only does one at a time? Do I understand this correctly?

File: files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md [hunk 4]
```diff
@@ -7,7 +7,9 @@ browser-compat: webextensions.api.storage.StorageArea.onChanged
...
+
+> **Note:** In Firefox, the information returned includes all keys within the storage area. Also, receiving a callback is possible when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md:12]:
```suggestion
> **Note:** In Firefox, the information returned includes all keys within the storage area. Also, the callback may be invoked when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md (hunk 3 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/api/storage/storagearea/onchanged/index.md:10]: @jpmedley, as I understand it, the difference is that Firefox returns all the keys from the storage area, whereas other browsers only return the changed ones only.

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 5]
```diff
@@ -7,9 +7,9 @@ browser-compat: webextensions.api.storage.onChanged
 
 {{AddonSidebar}}
 
-Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area.
+Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area, returning details of the changed keys. A callback is called only when there are changes to the underlying data.
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:10]:
```suggestion
Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, {{WebExtAPIRef('storage.StorageArea.remove','storageArea.remove')}}, or {{WebExtAPIRef('storage.StorageArea.clear','storageArea.clear')}} executes against a storage area, returning details of only changed keys. A callback is called only when there are changes to the underlying data.
```

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 6]
```diff
@@ -7,9 +7,9 @@ browser-compat: webextensions.api.storage.onChanged
...
 
-When this event is triggered by {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, it's possible to receive a callback when there is no change to the underlying data. Also, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. The extension can determine the changes that occurred by examining the content of the `changes` argument received by the `onChanged` listeners.
+> **Note:** In Firefox, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. Also, a callback may be invoked when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```
Comment (jpmedley) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:12]:
```suggestion
> **Note:** In Firefox, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against whether they changed or not. Also, a callback may be invoked when there is no change to the underlying data. Details of the changed items are found by examining each returned key's {{WebExtAPIRef('storage.StorageChange')}} object. See [Firefox bug 1833153](https://bugzil.la/1833153).
```

**Review by jpmedley — CHANGES_REQUESTED** [review]
Sorry for the delay. I had a long weekend. I don't usually request multiple passes of changes. I'm making an exception because I think we need to clarify the difference between Firefox and other browsers.

**Review by jpmedley — APPROVED** [review]

**Pull request merged**